tensorflow/go/attrs.go

```diff
@@ -0,0 +1,205 @@
+			if err := status.Err(); err != nil {
+				return nil, err
+			}
+			vals := make([]DataType, meta.list_size)
+			for i, val := range list {
+				vals[i] = DataType(val)
+			}
+			return vals, nil
+
+		case C.TF_ATTR_TENSOR:
+			list := make([]*C.TF_Tensor, meta.list_size)
+			C.TF_OperationGetAttrTensorList(op.c, cname, &list[0], C.int(meta.list_size), status.c)
+			if err := status.Err(); err != nil {
+				return nil, err
+			}
+			vals := make([]*Tensor, meta.list_size)
+			for i, t := range list {
+				vals[i] = newTensorFromC(t)
+			}
+			return vals, nil
+
+		case C.TF_ATTR_SHAPE:
+			dims := make([]*C.int64_t, meta.list_size)
+			numDims := make([]C.int, meta.list_size)
+			storage := make([]C.int64_t, meta.total_size)
+			C.TF_OperationGetAttrShapeList(op.c, cname, &dims[0], &numDims[0], C.int(meta.list_size), &storage[0], C.int(meta.total_size), status.c)
+			if err := status.Err(); err != nil {
+				return nil, err
+			}
+			list := make([]Shape, meta.list_size)
+			for i, dim := range dims {
+				numDim := numDims[i]
+				// A []C.int64_t slice backed by C memory.
+				// See: https://github.com/golang/go/wiki/cgo#turning-c-arrays-into-go-slices
+				slice := (*[1 << 30]C.int64_t)(unsafe.Pointer(dim))[:numDim:numDim]
+				s := Shape{
+					dims: make([]int64, numDim),
+				}
+				for i, dim := range slice {
+					s.dims[i] = int64(dim)
+				}
+				list[i] = s
+			}
+			return list, nil
+
+		default:
+			return nil, errors.Errorf("list type %v not supported yet", attrType)
+		}
+	}
+
+	switch attrType {
+	case C.TF_ATTR_STRING:
+		v := make([]C.char, meta.total_size)
+		C.TF_OperationGetAttrString(op.c, cname, unsafe.Pointer(&v[0]), C.size_t(meta.total_size), status.c)
+		if err := status.Err(); err != nil {
+			return nil, err
+		}
+		return C.GoStringN(&v[0], C.int(meta.total_size)), nil
+
+	case C.TF_ATTR_INT:
+		var v C.int64_t
+		C.TF_OperationGetAttrInt(op.c, cname, &v, status.c)
+		return int64(v), status.Err()
+
+	case C.TF_ATTR_FLOAT:
+		var v C.float
+		C.TF_OperationGetAttrFloat(op.c, cname, &v, status.c)
+		return float32(v), status.Err()
+
+	case C.TF_ATTR_BOOL:
+		var v C.uchar
+		C.TF_OperationGetAttrBool(op.c, cname, &v, status.c)
+		return v == 1, status.Err()
+
+	case C.TF_ATTR_TYPE:
+		var v C.TF_DataType
+		C.TF_OperationGetAttrType(op.c, cname, &v, status.c)
+		return DataType(v), status.Err()
+
+	case C.TF_ATTR_TENSOR:
+		var v *C.TF_Tensor
+		C.TF_OperationGetAttrTensor(op.c, cname, &v, status.c)
+		if err := status.Err(); err != nil {
+			return nil, err
+		}
+		return newTensorFromC(v), nil
+
+	case C.TF_ATTR_SHAPE:
+		numDims := meta.total_size
+		dims := make([]C.int64_t, numDims)
+		C.TF_OperationGetAttrShape(op.c, cname, (*C.int64_t)(unsafe.Pointer(&dims[0])), C.int(numDims), status.c)
+		if err := status.Err(); err != nil {
+			return nil, err
+		}
+		s := Shape{
+			dims: make([]int64, numDims),
+		}
+		for i, n := range dims {
```

Comment:
Optional: It may make sense to create a helper function:

```go
func makeShape(cdims []C.int64_t) Shape
```

That you can share between here and the list-valued shape construction.

Reply:
Done